docs/webhooks/example.go

```diff
@@ -0,0 +1,138 @@
+	currentVersion = "v1"
+	secret         = "tskey-webhook-xxxxx" // sensitive, here just as an example
+)
+
+var (
+	errNotSigned     = errors.New("webhook has no signature")
+	errInvalidHeader = errors.New("webhook has an invalid signature")
+)
+
+func main() {
+	http.HandleFunc("/webhook", webhooksHandler)
+	err := http.ListenAndServe(":80", nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func webhooksHandler(w http.ResponseWriter, req *http.Request) {
+	defer req.Body.Close()
+	events, err := verifyWebhookSignature(req, secret)
+	if err != nil {
+		log.Printf("error validating signature: %v\n", err)
+	} else {
+		log.Printf("events received %v\n", events)
+		// Do something with your events :)
+	}
+}
+
+// verifyWebhookSignature checks the request's "Tailscale-Webhook-Signature"
+// header to verify that the events were signed by your webhook secret.
+// If verification fails, an error is reported.
+// If verification succeeds, the list of contained events is reported.
+func verifyWebhookSignature(req *http.Request, secret string) (events []event, err error) {
+	defer req.Body.Close()
+
+	// Grab the signature sent on the request header.
+	timestamp, signatures, err := parseSignatureHeader(req.Header.Get("Tailscale-Webhook-Signature"))
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify that the timestamp is recent.
+	// Here, we use a threshold of 5 minutes.
+	if timestamp.Before(time.Now().Add(-time.Minute * 5)) {
+		return nil, fmt.Errorf("invalid header: timestamp older than 5 minutes")
+	}
+
+	// Form the expected signature.
+	b, err := io.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+	mac := hmac.New(sha256.New, []byte(secret))
+	mac.Write([]byte(fmt.Sprint(time.Now().Unix())))
+	mac.Write([]byte("."))
+	mac.Write(b)
+	want := hex.EncodeToString(mac.Sum(nil))
+
+	// Verify that the signatures match.
+	if signatures[currentVersion] != want {
+		return nil, fmt.Errorf("signature does not match: want = %q, got = %q", want, signatures[currentVersion])
+	}
+
+	// If verified, return the events.
+	if err := json.Unmarshal(b, &events); err != nil {
+		return nil, err
+	}
+	return events, nil
+}
+
+// parseSignatureHeader splits header into its timestamp and included signatures.
+// The signatures are reported as a map of version (e.g. "v1") to signature content.
+func parseSignatureHeader(header string) (timestamp time.Time, signatures map[string]string, err error) {
+	if header == "" {
+		return time.Time{}, nil, fmt.Errorf("request has no signature")
+	}
+
+	signatures = make(map[string]string)
+	pairs := strings.Split(header, ",")
+	for _, pair := range pairs {
+		parts := strings.Split(pair, "=")
+		if len(parts) != 2 {
+			return time.Time{}, nil, errNotSigned
+		}
+
+		switch parts[0] {
+		case "t":
+			tsint, err := strconv.ParseInt(parts[1], 10, 64)
+			if err != nil {
+				return time.Time{}, nil, errInvalidHeader
+			}
+			timestamp = time.Unix(tsint, 0)
+		case currentVersion:
+			signatures[parts[0]] = parts[1]
```

Comment:
This doesn't handle the case where there are multiple signatures of the same type... though I guess it also doesn't break and just takes the last one specified in the header. I don't know how much we want to optimize for, but perhaps a `map[string][]string` and then verifying in a loop above (on line 93)?

Reply:
Would we have multiple of the same type? Oh for rotation, they would both still be v1? And v2 only comes if there's a new secret format?

Reply:
Nice, ok updated!